File: app/src/main/java/com/tomclaw/drawa/draw/tools/Brush.java
package com.tomclaw.drawa.draw.tools;

import android.graphics.Paint;
import android.graphics.Path;

/**
 * Created by solkin on 17.03.17.
 */
//@EBean
public class Brush extends Tool {

    private static final float RADIUS_MULTIPLIER = 2;

    private int startX, startY;
    private int prevX, prevY;
    private Path path;

    Brush() {
    }

    @Override
    public void onInitialize() {
        this.path = new Path();
    }

    @Override
    Paint initPaint() {
        Paint paint = new Paint(Paint.ANTI_ALIAS_FLAG);
        paint.setAntiAlias(true);
        paint.setStyle(Paint.Style.STROKE);
        paint.setStrokeJoin(Paint.Join.ROUND);
        paint.setStrokeCap(Paint.Cap.ROUND);
        return paint;
    }

    @Override
    int getAlpha() {
        return 0xff;
    }

    @Override
    public void onTouchDown(int x, int y) {
        startX = x;
        startY = y;
        resetRadius();

        path.moveTo(x, y);
        path.lineTo(x, y);

        prevX = x;
        prevY = y;

        drawPath(path);
    }

    @Override
    public void onTouchMove(int x, int y) {
        if (path.isEmpty()) {
            path.moveTo(prevX, prevY);
        }
        path.quadTo(prevX, prevY, (x + prevX) / 2, (y + prevY) / 2);

        int deltaX = Math.abs(x - prevX);
        int deltaY = Math.abs(y - prevY);
        double length = Math.sqrt(deltaX * deltaX + deltaY * deltaY);
        int radius = getRadius();
        if (length < getBaseRadius() / 5) {
            radius += 2;

            path.reset();
            path.moveTo(prevX, prevY);
            path.lineTo(x, y);
        } else {
            radius -= 2;
        }
        if (radius > (getBaseRadius() / RADIUS_MULTIPLIER) && radius < (getBaseRadius() * RADIUS_MULTIPLIER)) {
            setRadius(radius);
        }

        prevX = x;
        prevY = y;

        drawPath(path);
    }

    @Override
    public void onTouchUp(int x, int y) {
        if (path.isEmpty()) {
            path.moveTo(prevX, prevY);
        }
        if (x == startX && y == startY) {
            path.lineTo(x + 0.1f, y);
        } else {
            path.lineTo(x, y);
        }

        drawPath(path);

        prevX = 0;
        prevY = 0;

        path.reset();
    }

    @Override
    public void onDraw() {
    }

    @Override
    public byte getType() {
        return TYPE_BRUSH;
    }
}

File: app/src/main/java/com/tomclaw/drawa/draw/tools/Brush.kt
package com.tomclaw.drawa.draw.tools

import android.graphics.Paint
import android.graphics.Path
import kotlin.math.abs
import kotlin.math.sqrt

class Brush : Tool() {

    private var startX: Int = 0
    private var startY: Int = 0
    private var prevX: Int = 0
    private var prevY: Int = 0
    private var path = Path()

    private var startStrokeSize: Float = 0.0f

    override val alpha = 0xff
    override val type = TYPE_BRUSH

    override fun initPaint() = Paint().apply {
        isAntiAlias = true
        isDither = true
        style = Paint.Style.STROKE
        strokeJoin = Paint.Join.ROUND
        strokeCap = Paint.Cap.ROUND
    }

    override fun onTouchDown(x: Int, y: Int) {
        resetRadius()
        startStrokeSize = strokeSize

        startX = x
        startY = y

        path.moveTo(x.toFloat(), y.toFloat())
        path.lineTo(x.toFloat(), y.toFloat())

        prevX = x
        prevY = y

        drawPath(path)
    }

    override fun onTouchMove(x: Int, y: Int) {
        if (path.isEmpty) {
            path.moveTo(prevX.toFloat(), prevY.toFloat())
        }
        path.quadTo(
                prevX.toFloat(),
                prevY.toFloat(),
                ((x + prevX).toFloat() / 2),
                ((y + prevY).toFloat() / 2)
        )

        val deltaX = abs(x - prevX)
        val deltaY = abs(y - prevY)
        val length = sqrt((deltaX * deltaX + deltaY * deltaY).toDouble())
        val sizeStep = defaultRadius / 20
        var size = strokeSize
        if (length * 5 < defaultRadius) {
            size += sizeStep

            path.reset()
            path.moveTo(prevX.toFloat(), prevY.toFloat())
            path.lineTo(x.toFloat(), y.toFloat())
        } else {
            size -= sizeStep
        }
        if (size > startStrokeSize / SIZE_MULTIPLIER && size < startStrokeSize * SIZE_MULTIPLIER) {
            strokeSize = size
        }

        prevX = x
        prevY = y

        drawPath(path)
    }

    override fun onTouchUp(x: Int, y: Int) {
        if (path.isEmpty) {
            path.moveTo(prevX.toFloat(), prevY.toFloat())
        }
        if (x == startX && y == startY) {
            path.lineTo(x + 0.1f, y.toFloat())
        } else {
            path.lineTo(x.toFloat(), y.toFloat())
        }

        drawPath(path)

        prevX = 0
        prevY = 0

        path.reset()
    }

    override fun onDraw() {}

}

private const val SIZE_MULTIPLIER = 2f
